processor/src/it/java/com/linecorp/decaton/processor/CoreFunctionalityTest.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright 2020 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.decaton.processor;
+
+import java.util.EnumSet;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+import org.junit.ClassRule;
+import org.junit.Test;
+
+import com.linecorp.decaton.testing.KafkaClusterRule;
+import com.linecorp.decaton.testing.processor.ProcessingGuarantee.GuaranteeType;
+import com.linecorp.decaton.testing.processor.ProcessorTestSuite;
+
+public class CoreFunctionalityTest {
+    @ClassRule
+    public static KafkaClusterRule rule = new KafkaClusterRule();
+
+    @Test(timeout = 30000)
+    public void testProcessConcurrent() {
+        ProcessorTestSuite
+                .builder(rule)
+                .configureProcessorsBuilder(builder -> builder.thenProcess((ctx, task) -> {
+                    // adding some pseudo random delay to generate out-of-order completion
+                    Thread.sleep(Math.abs(task.hashCode()) % 10);
+                }))
+                .propertySupplier(StaticPropertySupplier.of(
+                        Property.ofStatic(ProcessorProperties.CONFIG_PARTITION_CONCURRENCY, 16)
+                ))
+                .build()
+                .run();
+    }
+
+    @Test(timeout = 30000)
+    public void testAsyncTaskCompletion() {
+        ExecutorService executorService = Executors.newFixedThreadPool(16);
+        ProcessorTestSuite
+                .builder(rule)
+                .configureProcessorsBuilder(builder -> builder.thenProcess((ctx, task) -> {
+                    DeferredCompletion completion = ctx.deferCompletion();
+                    CompletableFuture.runAsync(() -> {
+                        try {
+                            Thread.sleep(Math.abs(task.hashCode()) % 10);
+                        } catch (InterruptedException e) {
+                            Thread.currentThread().interrupt();
+                            throw new RuntimeException(e);
+                        } finally {
+                            completion.complete();
+                        }
+                    }, executorService);
+                }))
+                // If task is completed asynchronously there's no guarantee about
+                // completion ordering nor serial processing
+                .semantics(EnumSet.of(
```

Comment:
Yeah I also considered about that, but I think there might be some semantics which is not included in default, but be enabled in "opt-in" manner (e.g. `at most once`)

To support such semantics, we'll add `includeSemantics`, and it's not so different from custom built-list ?

Reply:
Hmm, I see.
As we talked before, I was thinking this "default it framework" as a shared function to test decaton's core/basic functionality that are considered to be applied for all instances (except few exceptions).
So that while
- no message loss
- ordering preserved
- serial processing
are reasonable to be applied for all the cases unless user configures it to opt-ted out trading it off with something else. On the other hand it sounds bit wired to me that this framework has a built-in check for at-most-once because it's a "feature" outside of core functionality that users configure with trade off with potential message loss.
So I feel like at-most-once check should be defined and manually performed in the class "AtMostOnceTest".
However, I just came up with that even for testing that we'll need some parts of this framework anyway - perform restart, rebalance to simulate some "basic events" that likely happen on typical deployment and some information collection as its interface already defined. 
So what about following approach?
- Still keep the current semantics list as required default
- Add a interface to allow users (it authors) inject custom impl of semantics check that is outside of framework built-in but still gets ran in the same way with built-in checks.

Reply:
I see, makes sense.
> So I feel like at-most-once check should be defined and manually performed in the class "AtMostOnceTest"

Sounds nice.
> Add a interface to allow users (it authors) inject custom impl

like this?
```java
ProcessingGuarantee atMostOnce = new AtMostOnce();

ProcessorTestSuite.builder(rule)
                  .excludeSemantics(AT_LEAST_ONCE_DELIVERY)
                  .customSemantics(atMostOnce)
...
```

Reply:
ya, that's what I'm expecting.